### DISCUSSION TITLE
utext and HTML escaping some parameters

### DISCUSSION THREAD

**Discussion by jameskennard** [thymeleaf/thymeleaf#891] [Q&A]
Hi,

I'm using message properties that contain some basic HTML markup and parameters e.g. `some.message=Hello <strong>{0}</strong>`. To output this I then use utext so that I don't lose the HTML formatting, but this means that the parameters should be HTML escaped first (as per the docs - https://www.thymeleaf.org/doc/tutorials/3.0/usingthymeleaf.html#messages).

```
 <p th:utext="#{home.welcome(${session.user.name})}">
   Welcome to our grocery store, Sebastian Pepper!
 </p>
```
> Note that the use of th:utext here means that the formatted message will not be escaped. This example assumes that user.name is  already escaped.

Under the hood the the:text processor uses `HtmlEscape.escapeHtml4Xml` from a 3rd party library. You can get to this easily by using `#strings.escapeXml` (which despite the name is using `HtmlEscape.escapeHtml4Xml`). I appreciate that the HTML is XML encoded, but it seems a bit odd to ask for XML escaping in an HTML template and there is no `#strings.escapeHtml`.

So I've done this:

```
<div th:utext="#{some.message(${#strings.escapeXml(someModel.name)})}" >...
```

Is there a best-practice way of escaping these parameters?

Thanks

**Comment by conorheffron** [comment]
@jameskennard 

I think you found the answer, however, I found a useful utility to replace characters with empty string in the following fashion: `${#strings.replace(applicationVersion, 'Version: ', '')`.
- This might be another option depending on the incoming characters for certain edge cases.

## Sample Code & Snippet:
 - https://github.com/conorheffron/ironoc-db/blob/main/src/main/resources/templates/navbar.html
```
<a class="navbar-text version-badge ml-3"
               th:href="'https://github.com/conorheffron/ironoc-db/releases/tag/v' + ${#strings.replace(applicationVersion, 'Version: ', '')}"
               target="_blank"
               th:text="${applicationVersion}"></a>
```

---

## For more recent information (which I have not tested out entirely), see below:
 - In Thymeleaf, by default, all text output is HTML-escaped to prevent XSS (cross-site scripting) attacks.
If you want to render raw markup (HTML, SVG, etc.) inside a template without escaping, you need to use the unescaped output features.

## 1. Using th:utext (Unescaped Text)
```
<p th:utext="${htmlContent}"></p>
th:text → escapes HTML (safe by default)
th:utext → outputs raw HTML without escaping
```
#### Example:
```
model.addAttribute("htmlContent", "<strong>Bold Text</strong>");
```
##### Rendered Output:
```
<p><strong>Bold Text</strong></p>
```
## 2. Using Inlined Expressions with [(...)] vs. [[...]]
 - When using inline expressions in Thymeleaf:
```
[[...]] → escaped output
[(${...})] → unescaped output
```
#### Example:
```
<p>Escaped: [[${htmlContent}]]</p>
<p>Unescaped: [(${htmlContent})]</p>
```
## 3. In TEXT Template Mode
 - If you are working in TEXT mode (not HTML/XML), you can use the unescaped variant:
```
[(${variable})]
```
 - This will insert the raw text without escaping.

---

## ⚠ Security Warning
#### Only use unescaped output (th:utext or [(${...})]) when:
 - The HTML content is trusted (e.g., generated by your application, not user input).
 - You have sanitized any user-provided HTML to prevent XSS.

## ✅ Best Practice:
 - If you must display HTML from user input, sanitize it first using libraries like:
1. OWASP Java HTML Sanitizer
2. Jsoup (Jsoup.clean(...))
